galley/pkg/fs/fssource_test.go

```diff
@@ -326,23 +328,63 @@ func TestFsSource_DeletePartResorceInFile(t *testing.T) {
 	if err != nil {
 		t.Fatalf("Unexpected error: %v", err)
 	}
-
-	log := logChannelOutput(ch, 4)
+	donec := make(chan bool)
 	expected := "[Event](Deleted: [VKey](type.googleapis.com/istio.policy.v1beta1.Rule:some.mixer.rule @v0))"
+	go checkEventOccurs(expected, ch, donec)
+	select {
+	case <-time.After(5 * time.Second):
+		t.Fatalf("Expected Event does not occur:\n%s\n", expected)
+	case <-donec:
+		return
+	}
+	s.Stop()
+}
 
-	if log != expected {
-		t.Fatalf("Event mismatch:\nActual:\n%s\nExpected:\n%s\n", log, expected)
+func TestFsSource_publishEvent(t *testing.T) {
+	fst = &fsTestSourceState{
+		ConfigFiles: map[string][]byte{"virtual_service.yml": []byte(virtualServiceYAML)},
+	}
+	fst.testSetup(t)
+	defer fst.testTeardown(t)
+	s, err := New(fst.rootPath)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
 	}
 
-	s.Stop()
+	if s == nil {
+		t.Fatal("Expected non nil source")
+	}
+	d := runtime.NewInMemoryDistributor()
+	processor := runtime.NewProcessor(s, d)
+	err = processor.Start()
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
 
+	// Wait "long enough"- the default publishing timing is 1 second
+	time.Sleep(time.Second * 2)
```

Comment:
Let me rephrase that: Do not use sleeps without any feedback loop. It is ok to use a sleep/retry pattern to coordinate, but a single sleep followed by a pass/fail lookup is recipe for flakiness.

You can do a sleep/retry, or (even better) you can add some sort of channel/publishing support to the in-memory distributor, so that you can wait on distribution. e.g.:
```go
func (s* InMemorySnapshot) ListenChanges(chan cancel, ListenerFn fn) {
   s.listenerMutex.Lock()
   snapshotChan := make(chan Snapshot, 1)
   s.listeners = append(s.listeners, listenerEntry{fn, snapshotChan})
   s.listenerMutex.Unlock()
   for {
     select {
       case <-cancel:
         close(snapshotChan)
         return
       case s <- snapshotChan:
          fn(s)
    }
  }
}
// Called internally when a snapshot is set
func (s *InMemorySnapshot)  notifyListeners() {
   s.listenerMutex.Lock()
   defer s.listenerMutex.Unlock()
   for _, e := range s.listeners {
      e.snapshotChan <= s.snapshot
   }
}
```

Reply:
Thanks Oz, fixed now.